control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -374,14 +375,21 @@ func (r *HostedControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.R
 		r.Log.Info("Finished reconciling hosted cluster version conditions")
 	}
 
-	if hostedControlPlane.Spec.KubeConfig != nil {
-		hostedControlPlane.Status.KubeConfig = hostedControlPlane.Spec.KubeConfig
+	kubeconfig := manifests.KASExternalKubeconfigSecret(hostedControlPlane.Namespace, hostedControlPlane.Spec.KubeConfig)
+	if err := r.Get(ctx, client.ObjectKeyFromObject(kubeconfig), kubeconfig); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
 	} else {
 		hostedControlPlane.Status.KubeConfig = &hyperv1.KubeconfigSecretRef{
-			Name: DefaultAdminKubeconfigName,
+			Name: kubeconfig.Name,
 			Key:  DefaultAdminKubeconfigKey,
 		}
+		if hostedControlPlane.Spec.KubeConfig != nil {
+			hostedControlPlane.Status.KubeConfig.Key = hostedControlPlane.Spec.KubeConfig.Key
+		}
 	}
+
 	hostedControlPlane.Status.Initialized = true
```

Comment:
Is it ok that this gets set to `true` even when .status.Kubeconfig might be nil?

Reply:
This is part of the CAPI interface and my understanding is that it is used by the CAPI controller to decide whether to start watching nodes, so it really should be set to true only when the kube apiserver is available and a kubeconfig is available for CAPI to use.
https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/machine_controller.go#L231-L236
/cc @enxebre